File: packages/ld-adapter/src/errors/ld-adapter-error.ts
export class LdAdapterError extends Error {
	constructor(
		message: string,
		readonly context?: Record<string, unknown>,
	) {
		super(message);
		this.name = 'LdAdapterError';
	}
}

export class UnresolvedVariationError extends LdAdapterError {
	constructor(message: string, context?: Record<string, unknown>) {
		super(message, context);
		this.name = 'UnresolvedVariationError';
	}
}

export class UnresolvedRuleError extends LdAdapterError {
	constructor(message: string, context?: Record<string, unknown>) {
		super(message, context);
		this.name = 'UnresolvedRuleError';
	}
}

export class LdRateLimitError extends LdAdapterError {
	constructor(
		message: string,
		readonly retryAfterMs?: number,
		context?: Record<string, unknown>,
	) {
		super(message, context);
		this.name = 'LdRateLimitError';
	}
}

export class LdApiError extends LdAdapterError {
	constructor(
		message: string,
		readonly status: number,
		readonly body?: unknown,
		context?: Record<string, unknown>,
	) {
		super(message, context);
		this.name = 'LdApiError';
	}
}

export class ApprovalRequiredError extends LdAdapterError {
	constructor(
		message: string,
		readonly environmentKey?: string,
		context?: Record<string, unknown>,
	) {
		super(message, context);
		this.name = 'ApprovalRequiredError';
	}
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/errors/LdAdapterError.kt
package com.ffpromo.ldadapter.errors

open class LdAdapterError(
    message: String,
    open val context: Map<String, Any?>? = null,
) : Exception(message)

class UnresolvedVariationError(
    message: String,
    context: Map<String, Any?>? = null,
) : LdAdapterError(message, context)

class UnresolvedRuleError(
    message: String,
    context: Map<String, Any?>? = null,
) : LdAdapterError(message, context)

class LdRateLimitError(
    message: String,
    val retryAfterMs: Long? = null,
    context: Map<String, Any?>? = null,
) : LdAdapterError(message, context)

class LdApiError(
    message: String,
    val status: Int,
    val body: Any? = null,
    context: Map<String, Any?>? = null,
) : LdAdapterError(message, context)

class ApprovalRequiredError(
    message: String,
    val environmentKey: String? = null,
    context: Map<String, Any?>? = null,
) : LdAdapterError(message, context)

class HttpResponseError(
    message: String,
    val status: Int,
    val responseBody: Any? = null,
    val headers: Map<String, String> = emptyMap(),
) : Exception(message)
